runtime/array.c

```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
   }
   return res;
 }
+
+CAMLprim value caml_array_fill (value array, value v_ofs, value v_len, value val)
+{
+  value old;
+  intnat ofs = Long_val(v_ofs);
+  intnat len = Long_val(v_len);
+  value* fp;
+
+  /* See caml_modify for invariants */
+
+#ifdef FLAT_FLOAT_ARRAY
+  if (Tag_val(array) == Double_array_tag) {
+    double d = Double_val (val);
+    for (; len > 0; len--, ofs++)
+      Store_double_flat_field(array, ofs, d);
+    return Val_unit;
+  }
+#endif
+  fp = &Field(array, ofs);
+  if (Is_young(array)) {
+    for (; len > 0; len--, fp++) *fp = val;
+  } else {
```

Comment:
The other cases (flat float arrays and minor arrays) are obviously correct and look like clear wins. However, this `else` block duplicates a lot of subtle code from `caml_modify` and is quite hard to follow. (It's no worse than `caml_modify` in this respect, but at least that one is commented).

I think the block could be replaced by the following, and would still optimise the cases advertised in the PR description:

```
for (; len > 0; len--, fp++) {
  if (*fp != val) caml_modify(fp, val);
}
```

Reply:
The optimizations mentioned in the description (avoiding function calls and repeated `Is_*` tests) apply also the major heap case.

I tested with the following:
````
let () =
  let a = Array.make 128 None in
  Gc.compact ();
  for i = 0 to 10000000 do
    Array.fill a 0 128 (Some i);
  done
````

The simplified version you suggests make this run 2.7x slower then the current PR.  I think we can live the the duplicated logic, considering the impact on performances and the size of the duplicated code (but adding mutual references between the two functions is certainly useful).  Do you agree?

Reply:
@stedolan : can you confirm if you are ok with the extra bit of code duplication, considering the observed gains?

Reply:
Thanks for the benchmark!

If you think this is an important case to optimise, then I agree this is the right way to do it. Your benchmark shows that you can't get the same level of performance by calling `caml_modify`, and the mutual references in comments give enough pointers for someone else to understand what's going on.

I still can't think of a program that would hit this case a lot, though. (The motivating example of Hashtbl.clear never does). Do you have some program in mind that would be sped up by this, or is this optimisation more for completeness?

Reply:
No, honestly, I don't have a program in mind that benefits from this specific case, but it reduces the performance "cliff" between the fast path and the other cases; and addressing it now reduces the administrative overhead of doing it later if it turns out to be useful.

Reply:
OK, makes sense to me!